packages/server/src/artifact.rs

```diff
@@ -16,99 +16,122 @@ impl Server {
+		&self,
+		path: &tg::Path,
+		txn: &database::Transaction<'_>,
+	) -> Result<tg::artifact::Id> {
 		// Get the metadata for the file system object at the path.
-		let metadata = tokio::fs::symlink_metadata(path)
+		let metadata = tokio::fs::symlink_metadata(&path)
 			.await
 			.wrap_err_with(|| format!(r#"Failed to get the metadata for the path "{path}"."#))?;
 
 		// Call the appropriate function for the file system object at the path.
-		let id = if metadata.is_dir() {
-			self.check_in_directory(path, &metadata)
+		if metadata.is_dir() {
+			self.check_in_directory(path, &metadata, txn)
 				.await
-				.wrap_err_with(|| {
-					format!(r#"Failed to check in the directory at path "{path}"."#)
-				})?
+				.wrap_err_with(|| format!(r#"Failed to check in the directory at path "{path}"."#))
 		} else if metadata.is_file() {
-			self.check_in_file(path, &metadata)
+			self.check_in_file(path, &metadata, txn)
 				.await
-				.wrap_err_with(|| format!(r#"Failed to check in the file at path "{path}"."#))?
+				.wrap_err_with(|| format!(r#"Failed to check in the file at path "{path}"."#))
 		} else if metadata.is_symlink() {
-			self.check_in_symlink(path, &metadata)
+			self.check_in_symlink(path, &metadata, txn)
 				.await
-				.wrap_err_with(|| format!(r#"Failed to check in the symlink at path "{path}"."#))?
+				.wrap_err_with(|| format!(r#"Failed to check in the symlink at path "{path}"."#))
 		} else {
-			return Err(error!(
+			Err(error!(
 				"The path must point to a directory, file, or symlink."
-			));
-		};
-
-		let output = tg::artifact::CheckInOutput { id };
-
-		Ok(output)
+			))
+		}
 	}
 
-	#[async_recursion]
-	async fn check_in_directory(
-		&self,
-		path: &tg::Path,
-		_metadata: &std::fs::Metadata,
-	) -> Result<tg::artifact::Id> {
-		// Read the contents of the directory.
-		let names = {
-			let _permit = self.file_descriptor_semaphore().acquire().await;
-			let mut read_dir = tokio::fs::read_dir(path)
-				.await
-				.wrap_err("Failed to read the directory.")?;
-			let mut names = Vec::new();
-			while let Some(entry) = read_dir
-				.next_entry()
-				.await
-				.wrap_err("Failed to get the directory entry.")?
-			{
-				let name = entry
-					.file_name()
-					.to_str()
-					.wrap_err("All file names must be valid UTF-8.")?
-					.to_owned();
-				names.push(name);
-			}
-			names
-		};
-
-		// Recurse into the directory's entries.
-		let entries = names
-			.into_iter()
-			.map(|name| async {
-				let path = path.clone().join(&name);
-				let arg = tg::artifact::CheckInArg { path };
-				let output = self.check_in_artifact(arg).await?;
-				let artifact = tg::Artifact::with_id(output.id);
-				Ok::<_, Error>((name, artifact))
-			})
-			.collect::<FuturesUnordered<_>>()
-			.try_collect()
-			.await?;
+	fn check_in_directory<'a>(
+		&'a self,
+		path: &'a tg::Path,
+		_metadata: &'a std::fs::Metadata,
+		txn: &'a database::Transaction<'_>,
+	) -> LocalBoxFuture<'a, Result<tg::artifact::Id>> {
```

Comment:
No, need to get rid of async-recursion, you can use `#[async_recursion(?Send)]`.